schemas/2015-08-01/Microsoft.Cache.json

```diff
@@ -1,173 +1,210 @@
+          "oneOf": [
+            {
+              "type": "object",
+              "additionalProperties": {
+                "type": "string"
+              },
+              "properties": {}
             },
-            "redisConfiguration": {
-              "oneOf": [
-                { 
-                  "type": "object"
-                },
-                {
-                  "$ref": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/expression"
-                }
-              ],
-              "description": "Microsoft.Cache/Redis/redisConfiguration"
+            {
+              "$ref": "https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression"
+            }
+          ],
+          "description": "All Redis Settings. Few possible keys: rdb-backup-enabled,rdb-storage-connection-string,rdb-backup-frequency,maxmemory-delta,maxmemory-policy,notify-keyspace-events,maxmemory-samples,slowlog-log-slower-than,slowlog-max-len,list-max-ziplist-entries,list-max-ziplist-value,hash-max-ziplist-entries,hash-max-ziplist-value,set-max-intset-entries,zset-max-ziplist-entries,zset-max-ziplist-value etc."
+        },
+        "enableNonSslPort": {
+          "oneOf": [
+            {
+              "type": "boolean"
             },
-            "virtualNetwork": {
-              "oneOf": [
-                {
-                  "$ref": "#/definitions/virtualNetworkId"
-                },
-                {
-                  "$ref": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#/definitions/expression"
-                }
-              ],
-              "description": "Microsoft.Cache/Redis (optional) the ARM resource ID of a Classic Virtual Network for the redis cache to join"
+            {
+              "$ref": "https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression"
+            }
+          ],
+          "description": "If the value is true, then the non-SLL Redis server port (6379) will be enabled."
+        },
+        "tenantSettings": {
+          "oneOf": [
+            {
+              "type": "object",
+              "additionalProperties": {
+                "type": "string"
+              },
+              "properties": {}
+            },
+            {
+              "$ref": "https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression"
+            }
+          ],
+          "description": "tenantSettings"
+        },
+        "shardCount": {
+          "oneOf": [
+            {
+              "type": "integer"
+            },
+            {
+              "$ref": "https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression"
+            }
+          ],
+          "description": "The number of shards to be created on a Premium Cluster Cache."
+        },
+        "virtualNetwork": {
+          "type": "string",
```

Comment:
Is this change because it is a better practice to leave these as type 'string' rather than introduce our own validation regex types for VirtualNetworkId and IP Address?

Reply:
I believe it may be that the [autorest extension](https://github.com/Azure/autorest.azureresourceschema) is trying to take a shortcut and reduce schema size when taking a union of two patterned strings (rather than generating a oneOf with the ARM expression, it's simply allowing any old string). In this case however, it feels like it would be more useful to keep the pattern. I'll look into this in more detail.

Reply:
Ah, the explanation is much simpler. This version (2015-08-01) of your swagger spec doesn't have a pattern attached to the staticIP property. I see that in later versions, it does, and the schema generator includes the pattern, for example:
https://github.com/Azure/azure-resource-manager-schemas/blob/bbc0c1c3a73cd4674524c5b276c42e421a3ed865/schemas/2018-03-01/Microsoft.Cache.json#L319-L330

Reply:
Ah, the explanation is much simpler. This version (2015-08-01) of your swagger spec doesn't have a pattern attached to the staticIP property. I see that in later versions, it does, and the schema generator includes the pattern, for example:
https://github.com/Azure/azure-resource-manager-schemas/blob/bbc0c1c3a73cd4674524c5b276c42e421a3ed865/schemas/2018-03-01/Microsoft.Cache.json#L319-L330

For any swagger-defined string properties without a pattern attached, the extension simply leaves them as-is, as an ARM expression is just a more restrictive string, so will already be covered. This logic is defined [here](https://github.com/Azure/autorest.azureresourceschema/blob/137c09e0404f6689e402285e8990e2c8c71eab12/src/ResourceSchemaWriter.cs#L121).